jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-metadata.js

```diff
@@ -17,160 +17,149 @@ define([
 ) {
     'use strict';
 
-    var CELL_PROPERTIES = ['col','row','width','height'];
+    var SPEC_VERSION = 1;
+    var GRID_DEFAULT = 'grid_default';
+    var REPORT_DEFAULT = 'report_default';
 
-    // must match `dashboard-actions` auth states
-    var DASHBOARD_LAYOUT = Object.freeze({
+    // must match `dashboard-actions.js` auth states
+    var DASHBOARD_VIEW = Object.freeze({
         GRID: 'grid',
         REPORT: 'report'
     });
 
-    // pull cell metadata from element data
-    function _getCellMetadata($elem) {
+    // Mapping of view type to view id
+    // This will change when multiple views of the same type are supported.
+    var VIEW_TO_ID = {
+        grid: GRID_DEFAULT,
+        report: REPORT_DEFAULT
+    };
+
+    function _getActiveView() {
+        var metadata = _getDashboardMetadata();
+        if (metadata) {
+            return metadata.activeView;
+        }
+    }
+
+    // get cell view metadata from element data
+    function _getCellViewMetadata($elem, viewId) {
+        // use active view if not specified
+        viewId = viewId || _getActiveView();
         if (!$elem.is('.cell')) {
             $elem = $elem.parents('.cell').first();
         }
         if ($elem.length > 0) {
             var metadata = $elem.data('cell').metadata;
-            return metadata &&
-                   metadata.urth &&
-                   metadata.urth.dashboard;
+            return metadata.extensions.jupyter_dashboards.views[viewId];
         }
     }
 
-    function _createEmptyUrthMetadata(cell, preserveExistingMetadata) {
+    function _ensureHas(obj, prop, defaultValue) {
+        defaultValue = defaultValue || {};
+        if (!obj.hasOwnProperty(prop)) {
+            obj[prop] = defaultValue;
+        }
+    }
+
+    function _createDashboardMetadata(cell) {
         var metadata = IPython.notebook.metadata;
         if (cell) {
             metadata = $(cell).data('cell').metadata;
         }
-        if (preserveExistingMetadata) {
-            metadata.urth = metadata.urth || {};
-            metadata.urth.dashboard = metadata.urth.dashboard || {};
-        } else {
-            metadata.urth = {
-                dashboard: {}
-            };
-        }
-        return metadata;
-    }
 
-    function _getDashboardLayout() {
-        var metadata = _getDashboardMetadata();
-        var layout = null;
-        if (metadata && metadata.layout) {
-            layout = metadata.layout;
+        // common metadata
+        _ensureHas(metadata, 'extensions');
+        _ensureHas(metadata.extensions, 'jupyter_dashboards');
```

Comment:
Why not update `_ensureHas` to take the full dot-delimited string, similar to what [dojo.getObject](https://dojotoolkit.org/reference-guide/1.7/dojo/getObject.html) does?  i.e. `_ensureHas(metadata, 'extensions.jupyter_dashboards')`.